File: ios/Classes/SocialSharePlugin.m
#import "SocialSharePlugin.h"
#import "constants/ShareErrorCode.h"
#import "PlatformShareFactory.h"
#import "util/PlatformUtil.h"
#import "PlatformShare.h"
#import "constants/ShareTypeConst.h"
#import "WXApi.h"

@implementation SocialSharePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
        methodChannelWithName:@"social_share_plugin"
              binaryMessenger:[registrar messenger]];
    SocialSharePlugin* instance = [[SocialSharePlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
    [registrar addApplicationDelegate:instance];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"registerPlatforms" isEqualToString:call.method]) {
      [PlatformShareFactory registerPlatform:call.arguments];
  } else if ([@"isClientInstalled" isEqualToString:call.method]) {
      NSNumber* platform = call.arguments[@"platform"];
      [PlatformUtil isClientInstalled:platform.intValue result:result];
  } else if ([@"share" isEqualToString:call.method]) {
      [self shareToPlatform:call.arguments result:result];
  } else {
    result(FlutterMethodNotImplemented);
  }

}

/**
 return nil if the key not found
 */
-(id)getDataFromDic:(NSDictionary*)dic key:(NSString*)key {
    if ([dic[key] isKindOfClass:[NSNull class]]) {
        return nil;
    }
    return dic[key];
}

-(void)shareToPlatform:(NSDictionary*)arguments result:(FlutterResult)result {
    NSNumber* contentType = [self getDataFromDic:arguments key:@"contentType"];
    NSNumber* platformId = [self getDataFromDic:arguments key:@"platform"];
    if (contentType == nil || platformId == nil) {
        [SocialSharePlugin resultFail:result errorMessage:nil];
        return;
    }
    NSString* title = [self getDataFromDic:arguments key:@"title"];
    NSString* text = [self getDataFromDic:arguments key:@"text"];
    NSString* imageFilePath = [self getDataFromDic:arguments key:@"imageFilePath"];
    NSString* webUrl = [self getDataFromDic:arguments key:@"webUrl"];
    NSObject<PlatformShare>* platformShare = [PlatformShareFactory getPlatformShare:platformId.intValue];
    if (platformShare == nil) {
        [SocialSharePlugin resultFail:result errorMessage:nil];
        return;
    }
    switch (contentType.intValue) {
        case TEXT: {
            if (text == nil) {
                [SocialSharePlugin resultFail:result errorMessage:@"text needed"];
                return;
            }
            [platformShare shareText:text result:result];
            break;
        }
        case IMAGE: {
            if (imageFilePath == nil) {
                [SocialSharePlugin resultFail:result errorMessage:@"image url needed"];
                return;
            }
            [platformShare shareImage:imageFilePath result:result];
            break;
        }
        case WEB_PAGE: {
            if (webUrl == nil) {
                [SocialSharePlugin resultFail:result errorMessage:@"webUrl needed"];
                return;
            }
            [platformShare shareWebpage:webUrl title:title desc:text imgFilePath:imageFilePath result:result];
            break;
        }
        default: {
            [SocialSharePlugin resultFail:result errorMessage:@"not support"];
            break;
        }
    }
}

+(NSDictionary*)generateResultDic:(BOOL)isSuccess errorCode:(NSString*)errorCode errorMessage:(NSString*)errorMessage {
    if (errorCode == nil) {
        return @{
            @"isShareSuccess":@(isSuccess),
            };
    } else if (errorMessage == nil) {
        return @{
            @"isShareSuccess":@(isSuccess),
            @"errorCode":errorCode,
            };
    } else {
        return @{
            @"isShareSuccess":@(isSuccess),
            @"errorCode":errorCode,
            @"errorMessage":errorMessage,
            };
    }
}

+(void)resultFail:(FlutterResult)result errorMessage:(NSString*)errorMessage {
    [self resultFail:result errorCode:ERROR_BEFORE_JUMP errorMessage:errorMessage];
}

+(void)resultFail:(FlutterResult)result errorCode:(NSString*)errorCode errorMessage:(NSString*)errorMessage {
    result([self generateResultDic:false errorCode:errorCode errorMessage:errorMessage]);
}

+(void)resultSuccess:(FlutterResult)result {
    result([self generateResultDic:true errorCode:nil errorMessage:nil]);
}

-(BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray * _Nonnull))restorationHandler {
    return [WXApi handleOpenUniversalLink:userActivity
    delegate:self];
}
@end

File: android/src/main/java/com/chejdj/social_share/SocialSharePlugin.java
package com.chejdj.social_share;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.chejdj.social_share.constants.ShareErrorCode;
import com.chejdj.social_share.constants.ShareTypeConst;
import com.chejdj.social_share.util.PlatformUtil;
import com.chejdj.social_share.util.Utils;

import java.io.File;
import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/** SocialSharePlugin */
public class SocialSharePlugin implements FlutterPlugin, MethodCallHandler,ActivityAware {
  private Activity activity;
  private BinaryMessenger messenger;
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    messenger = flutterPluginBinding.getBinaryMessenger();
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    activity = null;
    messenger = null;
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
    activity = activityPluginBinding.getActivity();
    channel = new MethodChannel(messenger, "social_share_plugin");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
  }

  @Override
  public void onDetachedFromActivity() {
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "registerPlatforms":
        PlatformShareFactory.registerPlatform(activity, call.<Map<Integer, Map<String, String>>>arguments());
        break;
      case "isClientInstalled":
        PlatformUtil.isClientInstalled(activity, call.<Integer>argument("platform"), result);
        break;
      case "share":
        shareToPlatform(activity, call, result);
        break;
      default:
        result.notImplemented();
        break;
    }
  }

  private void shareToPlatform(Activity activity, MethodCall call, Result result) {
    Integer contentType = call.argument("contentType");
    Integer platformId = call.argument("platform");
    if (contentType == null || platformId == null) {
      resultFail(result, null);
      return;
    }
    String title = call.argument("title");
    String text = call.argument("text");
    String imageFilePath = call.argument("imageFilePath");
    String webUrl = call.argument("webUrl");
    String musicUrl = call.argument("musicUrl");
    String videoUrl = call.argument("videoUrl");
    String filePath = call.argument("filePath");
    String pkgName = call.argument("pkgNameAndroid");
    PlatformShare platformShare = PlatformShareFactory.getPlatformShare(platformId, pkgName);
    File imageFile = Utils.getFileWithPath(imageFilePath);
    File file = Utils.getFileWithPath(filePath);
    if (platformShare == null) {
      resultFail(result, null);
      return;
    }
    switch (contentType) {
      case ShareTypeConst.TEXT:
        platformShare.shareText(text, result, activity);
        break;
      case ShareTypeConst.IMAGE: {
        if (imageFile == null) {
          resultFail(result, "image url needed");
          break;
        }
        platformShare.shareImage(activity, imageFile, result);
        break;
      }
      case ShareTypeConst.WEB_PAGE: {
        if (Utils.isStringNullOrEmpty(webUrl)) {
          resultFail(result, "webUrl needed");
          break;
        }
        platformShare.shareWebpage(webUrl, title, text, imageFile, result, activity);
        break;
      }
      case ShareTypeConst.MUSIC: {
        if (Utils.isStringNullOrEmpty(musicUrl)) {
          resultFail(result, "musicUrl needed");
          break;
        }
        platformShare.shareMusic(musicUrl, title, text, imageFile, result, activity);
        break;
      }
      case ShareTypeConst.VIDEO: {
        if (Utils.isStringNullOrEmpty(videoUrl)) {
          resultFail(result, "musicUrl needed");
          break;
        }
        platformShare.shareVideo(videoUrl, title, text, imageFile, result, activity);
        break;
      }
      case ShareTypeConst.FILE: {
        if (file == null) {
          resultFail(result, "file url needed");
          break;
        }
        platformShare.shareFile(file, title, text, activity, result);
      }
      default: {
        break;
      }
    }
  }

  public static void resultFail(Result result, String errorMessage) {
    resultFail(result, ShareErrorCode.ERROR_BEFORE_JUMP, errorMessage);
  }

  public static void resultFail(Result result, String errorCode, String errorMessage) {
    result.success(generateResultMap(false, errorCode, errorMessage));
  }

  public static void resultSuccess(Result result) {
    result.success(generateResultMap(true, null, null));
  }

  public static Map<String, Object> generateResultMap(boolean isSuccess, String errorCode, String errorMessage) {
    final Map<String, Object> map = new HashMap<>();
    map.put("isShareSuccess", isSuccess);
    map.put("errorCode", errorCode);
    map.put("errorMessage", errorMessage);
    return map;
  }
}
